sycl/source/detail/link_graph.hpp

```diff
@@ -0,0 +1,421 @@
+#include <vector>
+
+namespace sycl {
+inline namespace _V1 {
+namespace detail {
+
+/// Representation of the link graph.
+template <typename T> class LinkGraph {
+protected:
+  // Representation of a node in the graph.
+  struct Node {
+    Node(const T &Val) : Value(Val) {}
+
+    T Value;
+    bool Alive = true;
+    // Parents and children are represented by the corresponding offsets into
+    // the node list. Since nodes in a node list never move, this simplifies
+    // copying the lists.
+    std::vector<size_t> Parents, Children;
+  };
+
+public:
+  LinkGraph() = default;
+  LinkGraph(LinkGraph &&Other) = default;
+
+  LinkGraph(const std::vector<T> &Values,
+            const std::vector<std::vector<size_t>> &Links) {
+    assert(Values.size() == Links.size());
+
+    Nodes.reserve(Values.size());
+    for (const T &Value : Values)
+      Nodes.emplace_back(Value);
+
+    // Create links after the nodes have all been materialized.
+    for (size_t I = 0; I < Values.size(); ++I) {
+      Node &N = Nodes[I];
+      N.Children = Links[I];
+      for (size_t ChildIndex : N.Children)
+        Nodes[ChildIndex].Parents.push_back(I);
+    }
+  }
+
+  LinkGraph Clone() const { return LinkGraph{*this}; }
+
+  /// Poisons the nodes in the graph based on a poisoning condition. Poisoning
+  /// a node causes the node to be removed from the graph will recursively
+  /// poison the parents of the node, even if they do not satisfy the poisoning
+  /// condition.
+  /// \param PoisonCondFunc is the poisoning condition.
+  template <typename PoisonCondFuncT>
+  void Poison(const PoisonCondFuncT &PoisonCondFunc) {
+    // Empty graph, so no work to be done.
+    if (Nodes.empty())
+      return;
+
+    // Keep track of seen nodes and how often they have been seen.
+    std::vector<uint32_t> Seen(Nodes.size(), 0);
+
+    // Keep a growing work-list of nodes to check. We should get through all
+    // nodes this way, excluding dead nodes.
+    std::vector<size_t> WorkList;
+    WorkList.reserve(Nodes.size());
+
+    // Pass through nodes to count dead nodes and add all leaves to the
+    // work-list.
+    size_t InitialDeadNodeCount = 0;
+    for (size_t I = 0; I < Nodes.size(); ++I) {
+      const Node &CurrentNode = Nodes[I];
+      if (!CurrentNode.Alive) {
+        Seen[I] = 1;
+        ++InitialDeadNodeCount;
+      } else if (CurrentNode.Children.empty()) {
+        Seen[I] = 1;
+        WorkList.push_back(I);
+      }
+    }
+
+    // Poison all leaves.
+    PoisonInner(PoisonCondFunc, WorkList, Seen);
+
+    // If the worklist doesn't have all live nodes, there must be cycles in the
+    // list. We handle these by finding any part of them that we haven't yet
+    // seen and process them in separation until there is no more work to do.
+    while (WorkList.size() < Nodes.size() - InitialDeadNodeCount) {
+      const size_t PreviousWorkListSize = WorkList.size();
+
+      // Find the first unseen node. This will be the basis of our new work.
```

Comment:
It looks like all unseen modes are added to the work list, not only the first one. Is this a mistake or intentional?

Reply:
Ah, good catch! That was definitely not intended.